fs/zbd_zenfs.cc

```diff
@@ -917,6 +918,91 @@ Zone *ZonedBlockDevice::AllocateZone(Env::WriteLifeTimeHint file_lifetime, bool
   return allocated_zone;
 }
 
+Zone *ZonedBlockDevice::AllocateDataZone(Env::WriteLifeTimeHint file_lifetime, bool is_wal) {
+  // Ignore lifetime hint for now
+  (void)file_lifetime;
+
+  Zone* ret = nullptr;
+  if (is_wal) {
+    fg_request_++;
+    do {
+      {
+        std::unique_lock<std::mutex> lk(active_zone_vec_mtx_);
+        ret = GetDataZoneLocked(0);  
+      }
+      if (ret == nullptr) {
+        // Resources not enough, release lock for background worker and waiting.
+        std::this_thread::yield();
+      }
+    } while (ret == nullptr);
+    fg_request_--;
+  } else {
+    // Normal file
+    int request = 0;
+    do {
+      // Test if there is high priority requsets.
+      request = fg_request_.load();
+      if (request == 0) {
+        // Fetch one when no contention occurs.
+        std::unique_lock<std::mutex> lk(active_zone_vec_mtx_);
+        ret = GetDataZoneLocked(reserved_active_zones);
+      } else {
+        std::this_thread::yield();
+      }
+    } while (ret == nullptr);
+  }
+  return ret;
+}
+
+Zone *ZonedBlockDevice::GetDataZoneLocked(int start) {
+  for (int i = start; i < active_zones_vec_.size(); ++i) {
+    if (active_zones_vec_[i]->state_ == kActive) {
+      active_zones_vec_[i]->state_ = kOccupied;
+      return active_zones_vec_[i];
+    }
+  }
+  return nullptr;
+}
+
+void ZonedBlockDevice::BgRefillActiveZones() {
+  data_worker_->SubmitJob(
+    [&]() {
+      for (int i = 0; i < active_zones_vec_.size(); ++i) {
+        if (active_zones_vec_[i]->state_ == kReadOnly) {
+          std::unique_lock<std::mutex> lk(active_zone_vec_mtx_);
+          BgReplaceFromIOZones(i);
+        }
+      }
+    });
+}
+
+void ZonedBlockDevice::BgReplaceFromIOZones(int index) {
+  std::unique_lock<std::mutex> lk(io_zones_mtx);
+  for (auto& z : io_zones) {
+    if (z->state_ == kEmpty) {
+      z->state_ = kActive;
+      z->open_for_write_ = true;
+      active_zones_vec_[index] = z;
+    }
+  }
+}
+
+void ZonedBlockDevice::BgResetDataZone(Zone* target) {
+
+  std::function<IOStatus(Zone*)> fn = [&](Zone* z) -> IOStatus {
+    return z->Reset();
```

Comment:
I prefer to handle errors inside this lambda function instead of using a separated error handler since we don't have a lot of different errors. And we don't have real error handlers (which means we could do nothing but print an error message when we meet an error......). But you can keep this logic since you've already did it.

Reply:
It's ok, I changed the background worker, now it supports both ways, lambda or outsider.